### PR TITLE
Fixes to various visualization problems

### PR DESCRIPTION
**Pull request by tomnelson**
Changes to be committed:

	modified:   jung-algorithms/pom.xml
	modified:   jung-algorithms/src/main/java/edu/uci/ics/jung/algorithms/layout/util/VisRunner.java
	new file:   jung-algorithms/src/main/resources/log4j.properties
	new file:   jung-algorithms/src/test/java/edu/uci/ics/jung/algorithms/layout/LayoutTest.java
	modified:   jung-samples/pom.xml
	modified:   jung-samples/src/main/java/edu/uci/ics/jung/samples/AddNodeDemo.java
	modified:   jung-samples/src/main/java/edu/uci/ics/jung/samples/AnimatingAddNodeDemo.java
	modified:   jung-samples/src/main/java/edu/uci/ics/jung/samples/BalloonLayoutDemo.java
	modified:   jung-samples/src/main/java/edu/uci/ics/jung/samples/ClusteringDemo.java
	modified:   jung-samples/src/main/java/edu/uci/ics/jung/samples/EdgeLabelDemo.java
	modified:   jung-samples/src/main/java/edu/uci/ics/jung/samples/GraphFromGraphMLDemo.java
	modified:   jung-samples/src/main/java/edu/uci/ics/jung/samples/GraphZoomScrollPaneDemo.java
	modified:   jung-samples/src/main/java/edu/uci/ics/jung/samples/L2RTreeLayoutDemo.java
	modified:   jung-samples/src/main/java/edu/uci/ics/jung/samples/LensDemo.java
	modified:   jung-samples/src/main/java/edu/uci/ics/jung/samples/PluggableRendererDemo.java
	modified:   jung-samples/src/main/java/edu/uci/ics/jung/samples/SatelliteViewDemo.java
	modified:   jung-samples/src/main/java/edu/uci/ics/jung/samples/ShortestPathDemo.java
	modified:   jung-samples/src/main/java/edu/uci/ics/jung/samples/ShowLayouts.java
	modified:   jung-samples/src/main/java/edu/uci/ics/jung/samples/SubLayoutDemo.java
	modified:   jung-samples/src/main/java/edu/uci/ics/jung/samples/TreeCollapseDemo.java
	modified:   jung-samples/src/main/java/edu/uci/ics/jung/samples/TreeLayoutDemo.java
	modified:   jung-samples/src/main/java/edu/uci/ics/jung/samples/VertexCollapseDemo.java
	modified:   jung-samples/src/main/java/edu/uci/ics/jung/samples/VertexCollapseDemoWithLayouts.java
	modified:   jung-samples/src/main/java/edu/uci/ics/jung/samples/VertexImageShaperDemo.java
	modified:   jung-samples/src/main/java/edu/uci/ics/jung/samples/VertexLabelAsShapeDemo.java
	modified:   jung-samples/src/main/java/edu/uci/ics/jung/samples/VisualizationImageServerDemo.java
	modified:   jung-samples/src/main/java/edu/uci/ics/jung/samples/WorldMapGraphDemo.java
	modified:   jung-visualization/pom.xml
	modified:   jung-visualization/src/main/java/edu/uci/ics/jung/visualization/BasicVisualizationServer.java
	modified:   jung-visualization/src/main/java/edu/uci/ics/jung/visualization/DefaultVisualizationModel.java
	modified:   jung-visualization/src/main/java/edu/uci/ics/jung/visualization/PluggableRenderContext.java
	modified:   jung-visualization/src/main/java/edu/uci/ics/jung/visualization/RenderContext.java
	modified:   jung-visualization/src/main/java/edu/uci/ics/jung/visualization/VisualizationModel.java
	modified:   jung-visualization/src/main/java/edu/uci/ics/jung/visualization/VisualizationServer.java
	modified:   jung-visualization/src/main/java/edu/uci/ics/jung/visualization/control/EditingModalGraphMouse.java
	modified:   jung-visualization/src/main/java/edu/uci/ics/jung/visualization/control/EditingPopupGraphMousePlugin.java
	modified:   jung-visualization/src/main/java/edu/uci/ics/jung/visualization/control/SatelliteVisualizationViewer.java
	modified:   jung-visualization/src/main/java/edu/uci/ics/jung/visualization/control/SimpleEdgeSupport.java
	modified:   jung-visualization/src/main/java/edu/uci/ics/jung/visualization/control/SimpleVertexSupport.java
	new file:   jung-visualization/src/main/java/edu/uci/ics/jung/visualization/decorators/AbstractEdgeShapeTransformer.java
	modified:   jung-visualization/src/main/java/edu/uci/ics/jung/visualization/decorators/EdgeShape.java
	modified:   jung-visualization/src/main/java/edu/uci/ics/jung/visualization/decorators/GradientEdgePaintTransformer.java
	modified:   jung-visualization/src/main/java/edu/uci/ics/jung/visualization/decorators/ParallelEdgeShapeTransformer.java
	modified:   jung-visualization/src/main/java/edu/uci/ics/jung/visualization/layout/BoundingRectangleCollector.java
	modified:   jung-visualization/src/main/java/edu/uci/ics/jung/visualization/layout/BoundingRectanglePaintable.java
	modified:   jung-visualization/src/main/java/edu/uci/ics/jung/visualization/layout/LayoutChangeListener.java
	modified:   jung-visualization/src/main/java/edu/uci/ics/jung/visualization/layout/LayoutEvent.java
	modified:   jung-visualization/src/main/java/edu/uci/ics/jung/visualization/layout/LayoutEventSupport.java
	modified:   jung-visualization/src/main/java/edu/uci/ics/jung/visualization/layout/LayoutTransition.java
	modified:   jung-visualization/src/main/java/edu/uci/ics/jung/visualization/layout/ObservableCachingLayout.java
	modified:   jung-visualization/src/main/java/edu/uci/ics/jung/visualization/picking/ClosestShapePickSupport.java
	modified:   jung-visualization/src/main/java/edu/uci/ics/jung/visualization/picking/LayoutLensShapePickSupport.java
	modified:   jung-visualization/src/main/java/edu/uci/ics/jung/visualization/picking/ShapePickSupport.java
	modified:   jung-visualization/src/main/java/edu/uci/ics/jung/visualization/picking/ViewLensShapePickSupport.java
	modified:   jung-visualization/src/main/java/edu/uci/ics/jung/visualization/renderers/BasicEdgeLabelRenderer.java
	modified:   jung-visualization/src/main/java/edu/uci/ics/jung/visualization/renderers/BasicEdgeRenderer.java
	modified:   jung-visualization/src/main/java/edu/uci/ics/jung/visualization/renderers/BasicRenderer.java
	modified:   jung-visualization/src/main/java/edu/uci/ics/jung/visualization/renderers/BasicVertexLabelRenderer.java
	modified:   jung-visualization/src/main/java/edu/uci/ics/jung/visualization/renderers/BasicVertexRenderer.java
	modified:   jung-visualization/src/main/java/edu/uci/ics/jung/visualization/renderers/CachingEdgeRenderer.java
	modified:   jung-visualization/src/main/java/edu/uci/ics/jung/visualization/renderers/CachingRenderer.java
	modified:   jung-visualization/src/main/java/edu/uci/ics/jung/visualization/renderers/CachingVertexRenderer.java
	modified:   jung-visualization/src/main/java/edu/uci/ics/jung/visualization/renderers/GradientVertexRenderer.java
	modified:   jung-visualization/src/main/java/edu/uci/ics/jung/visualization/renderers/Renderer.java
	modified:   jung-visualization/src/main/java/edu/uci/ics/jung/visualization/renderers/ReshapingEdgeRenderer.java
	modified:   jung-visualization/src/main/java/edu/uci/ics/jung/visualization/renderers/VertexLabelAsShapeRenderer.java
	new file:   jung-visualization/src/main/java/edu/uci/ics/jung/visualization/spatial/Spatial.java
	new file:   jung-visualization/src/main/java/edu/uci/ics/jung/visualization/spatial/SpatialGrid.java
	modified:   jung-visualization/src/main/java/edu/uci/ics/jung/visualization/subLayout/GraphCollapser.java
	modified:   jung-visualization/src/main/java/edu/uci/ics/jung/visualization/transform/shape/MagnifyImageLensSupport.java
	modified:   jung-visualization/src/main/java/edu/uci/ics/jung/visualization/transform/shape/ViewLensSupport.java
	new file:   jung-visualization/src/main/java/edu/uci/ics/jung/visualization/util/Context.java
	new file:   jung-visualization/src/main/java/edu/uci/ics/jung/visualization/util/LayoutMediator.java
	new file:   jung-visualization/src/main/resources/log4j.properties
	modified:   pom.xml

### CODE REVIEW

**Comment by jrtom** [comment]
Can you summarize at a medium-high level what changes are involved here?

Also, did you intend to commit the log4j.properties file?

**Comment by jbduncan** [comment]
The first thing that pops out to me is that log4j is being used for logging. Do we really need logging for JUNG? If we do really need logging, then I believe slf4j should at least be used as a frontend for log4j, to better allow users to swap out log4j for their preferred logging framework if they desire.

File: jung-algorithms/src/main/java/edu/uci/ics/jung/algorithms/layout/util/VisRunner.java [hunk 1]
```diff
@@ -10,6 +10,7 @@
 package edu.uci.ics.jung.algorithms.layout.util;
 
 import edu.uci.ics.jung.algorithms.util.IterativeContext;
+import org.apache.log4j.Logger;
```
Comment (jrtom) [inline jung-algorithms/src/main/java/edu/uci/ics/jung/algorithms/layout/util/VisRunner.java]: Please remove the logging code (here and elsewhere)--both the Logger stuff and the System.out.println()s.

**Comment by tomnelson** [comment]
I removed all changes to jung-algorithms.

**Comment by tomnelson** [comment]
- removed all the statefulness (mostly of Network) that broke many demos
- yes, i intended to add logging, but just to visualization and samples

**Comment by jrtom** [comment]
OK.  I will take a look at this as soon as I can, but I want to finish throwing together the design doc I mentioned in #104 first, and then I'd like to see those designs compared in that doc before I make a final call on this.

Thanks for putting this together, at any rate; I think that it will really help to be able to see how your ideas are working in action.

File: jung-visualization/src/main/java/edu/uci/ics/jung/visualization/BasicVisualizationServer.java [hunk 2]
```diff
@@ -224,12 +228,19 @@ public void setRenderer(Renderer<V, E> r) {
...
 
-  public void setGraphLayout(Layout<V> layout) {
+  public void setLayoutMediator(LayoutMediator<V, E> layoutMediator, Dimension d) {
+    model.setLayoutMediator(layoutMediator, d);
+  }
+
+  public void setLayoutMediator(LayoutMediator<V, E> layoutMediator) {
+    if (log.isDebugEnabled()) {
+      log.debug("setLayoutMediator to " + layoutMediator);
+    }
```
Comment (jbduncan) [inline jung-visualization/src/main/java/edu/uci/ics/jung/visualization/BasicVisualizationServer.java]: It's not really clear to me that a JUNG user would find this logging useful. After all, the typical developer has access to a debugger for cases like this, and if JUNG ends up used in a command-line application, we would clutter the output log, annoying users (who would experience it firsthand) and developers (who would have to find a way to control or disable it) alike.

File: jung-visualization/src/main/java/edu/uci/ics/jung/visualization/BasicVisualizationServer.java [hunk 3]
```diff
@@ -304,7 +315,48 @@ protected void renderGraph(Graphics2D g2d) {
...
+    spatialTransform.concatenate(
+        renderContext.getMultiLayerTransformer().getTransformer(Layer.VIEW).getTransform());
+    spatialTransform.concatenate(
+        renderContext.getMultiLayerTransformer().getTransformer(Layer.LAYOUT).getTransform());
+
+    try {
+      spatialTransform = spatialTransform.createInverse();
+    } catch (Exception e) {
+      e.printStackTrace();
+    }
```
Comment (jbduncan) [inline jung-visualization/src/main/java/edu/uci/ics/jung/visualization/BasicVisualizationServer.java]: I wonder if it would be better to let exceptions thrown here to propogate to the caller.

File: jung-visualization/src/main/java/edu/uci/ics/jung/visualization/BasicVisualizationServer.java [hunk 4]
```diff
@@ -304,7 +315,48 @@ protected void renderGraph(Graphics2D g2d) {
...
+      log.debug(
+          "layoutXform scale is "
+              + renderContext
+                  .getMultiLayerTransformer()
+                  .getTransformer(Layer.LAYOUT)
+                  .getTransform()
+                  .getScaleX());
+      log.debug("newXform scale is " + newXform.getScaleX());
+      log.debug("this thing is " + bigger.getBounds());
+    }
```
Comment (jbduncan) [inline jung-visualization/src/main/java/edu/uci/ics/jung/visualization/BasicVisualizationServer.java]: Again, it's not really clear to me why we're logging this debug information.

File: jung-visualization/pom.xml [hunk 5]
```diff
@@ -9,9 +9,33 @@
   <artifactId>jung-visualization</artifactId>
   <name>JUNG - Visualization Support</name>
   <description>Core visualization support for the JUNG project</description>
-
+  <properties>
+    <version.commons-logging>1.2</version.commons-logging>
+    <version.log4j>1.2.16</version.log4j>
```
Comment (jbduncan) [inline jung-visualization/pom.xml]: If we really do want to keep log4j (which I admit I'm sceptical about currently), please consider using log4j 2.

File: jung-visualization/src/main/java/edu/uci/ics/jung/visualization/BasicVisualizationServer.java [hunk 6]
```diff
@@ -321,8 +373,16 @@ protected void renderGraph(Graphics2D g2d) {
     if (layout instanceof Caching) {
       ((Caching) layout).clear();
     }
-
-    renderer.render();
+    log.debug(
+        "render nodes from "
+            + model.getLayoutMediator()
+            + " with nodes "
+            + model.getLayoutMediator().getNetwork().nodes());
```
Comment (jbduncan) [inline jung-visualization/src/main/java/edu/uci/ics/jung/visualization/BasicVisualizationServer.java]: Ditto.

File: jung-visualization/src/main/java/edu/uci/ics/jung/visualization/DefaultVisualizationModel.java [hunk 7]
```diff
@@ -65,37 +66,43 @@ public void stateChanged(ChangeEvent e) {
...
       viewSize = new Dimension(600, 600);
     }
-    Dimension layoutSize = layout.getSize();
+    Dimension layoutSize = newLayout.getSize();
     // if the layout has NOT been initialized yet, initialize its size
     // now to the size of the VisualizationViewer window
     if (layoutSize == null) {
-      layout.setSize(viewSize);
+      newLayout.setSize(viewSize);
+      //      newLayout.setSize(viewSize);
```
Comment (jbduncan) [inline jung-visualization/src/main/java/edu/uci/ics/jung/visualization/DefaultVisualizationModel.java]: I believe this comment has very little use. Would you please consider removing it.

File: jung-visualization/src/main/java/edu/uci/ics/jung/visualization/PluggableRenderContext.java [hunk 8]
```diff
@@ -105,11 +104,10 @@
   protected GraphicsDecorator graphicsContext;
 
   private EdgeShape<E> edgeShape;
-
+  //Function<Context<Network,E>, Shape>
```
Comment (jbduncan) [inline jung-visualization/src/main/java/edu/uci/ics/jung/visualization/PluggableRenderContext.java]: Ditto.

File: jung-visualization/src/main/java/edu/uci/ics/jung/visualization/PluggableRenderContext.java [hunk 9]
```diff
@@ -105,11 +104,10 @@
...
   PluggableRenderContext(Network<V, E> graph) {
-    this.network = graph;
-    this.edgeShape = new EdgeShape<E>(graph);
-    this.edgeShapeTransformer = edgeShape.new QuadCurve();
+    //    this.edgeShape = new EdgeShape<E>(graph);
```
Comment (jbduncan) [inline jung-visualization/src/main/java/edu/uci/ics/jung/visualization/PluggableRenderContext.java]: Ditto.

Please look back over all the other comments you added and consider whether they're really needed.

File: jung-visualization/src/main/java/edu/uci/ics/jung/visualization/control/EditingModalGraphMouse.java [hunk 10]
```diff
@@ -64,9 +62,9 @@ public EditingModalGraphMouse(
       float in,
       float out) {
     super(in, out);
-    Preconditions.checkArgument(
-        rc.getNetwork() instanceof MutableNetwork,
-        "Supplied Network instance must be a MutableNetwork");
+    //    Preconditions.checkArgument(
+    //            layoutMediator.getNetwork() instanceof MutableNetwork,
+    //        "Supplied Network instance must be a MutableNetwork");
```
Comment (jbduncan) [inline jung-visualization/src/main/java/edu/uci/ics/jung/visualization/control/EditingModalGraphMouse.java]: Please consider uncommenting these lines.

File: jung-visualization/src/main/java/edu/uci/ics/jung/visualization/VisualizationServer.java [hunk 11]
```diff
@@ -68,15 +67,15 @@
...
+  //   * Replaces the current graph layout with {@code layout}.
+  //   *
+  //   * @param layout the new layout to set
+  //   */
+  //  void setGraphLayout(Layout<V> layout);
 
-  /** @return the current graph layout. */
-  Layout<V> getGraphLayout();
+  //  /** @return the current graph layout. */
+  //  Layout<V> getGraphLayout();
```
Comment (jbduncan) [inline jung-visualization/src/main/java/edu/uci/ics/jung/visualization/VisualizationServer.java]: Please consider uncommenting or removing these lines.

File: jung-visualization/src/main/java/edu/uci/ics/jung/visualization/VisualizationServer.java [hunk 12]
```diff
@@ -68,15 +67,15 @@
...
-   *
-   * @param layout the new layout to set
-   */
-  void setGraphLayout(Layout<V> layout);
+  //  /**
+  //   * Replaces the current graph layout with {@code layout}.
+  //   *
+  //   * @param layout the new layout to set
+  //   */
+  //  void setGraphLayout(Layout<V> layout);
```
Comment (jbduncan) [inline jung-visualization/src/main/java/edu/uci/ics/jung/visualization/VisualizationServer.java]: Please consider uncommenting or removing these lines.

File: jung-visualization/src/main/java/edu/uci/ics/jung/visualization/decorators/AbstractEdgeShapeTransformer.java [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+/**
+ * An interface for decorators that return a <code>Shape</code> for a specified edge.
+ *
+ * @author Tom Nelson
+ */
+public abstract class AbstractEdgeShapeTransformer<V, E>
+    implements Function<Context<Network<V, E>, E>, Shape> {
+
+  /** Specifies how far apart to place the control points for edges being drawn in parallel. */
+  protected float control_offset_increment = 20.f;
```
Comment (jbduncan) [inline jung-visualization/src/main/java/edu/uci/ics/jung/visualization/decorators/AbstractEdgeShapeTransformer.java]: Please make this camelCase.

File: jung-visualization/src/main/java/edu/uci/ics/jung/visualization/layout/LayoutTransition.java [hunk 14]
```diff
@@ -1,52 +1,68 @@
...
   protected VisualizationViewer<V, E> vv;
 
   public LayoutTransition(
-      VisualizationViewer<V, E> vv, Layout<V> startLayout, Layout<V> endLayout) {
+      VisualizationViewer<V, E> vv,
+      LayoutMediator<V, E> startLayoutMediator,
+      LayoutMediator<V, E> endLayoutMediator) {
+    if (log.isDebugEnabled()) {
+      log.debug("transition from " + startLayoutMediator + " to " + endLayoutMediator);
+    }
```
Comment (jbduncan) [inline jung-visualization/src/main/java/edu/uci/ics/jung/visualization/layout/LayoutTransition.java]: Ditto regarding earlier logging comments.

Please consider whether logging statements are needed at all.

File: jung-visualization/src/main/java/edu/uci/ics/jung/visualization/renderers/GradientVertexRenderer.java [hunk 15]
```diff
@@ -66,17 +62,18 @@ public GradientVertexRenderer(
     this.pickedColorTwo = pickedColorTwo;
     this.pickedState = vv.getPickedVertexState();
     this.cyclic = cyclic;
-    this.layout = vv.getGraphLayout();
-    this.renderContext = vv.getRenderContext();
+    //    this.layout = vv.getGraphLayout();
+    //    this.renderContext = vv.getRenderContext();
```
Comment (jbduncan) [inline jung-visualization/src/main/java/edu/uci/ics/jung/visualization/renderers/GradientVertexRenderer.java]: Ditto regarding earlier comment on commenting-out code.

For all places where you've done this, please consider whether they really need to be commented out, and if instead they can be uncommented again or simply removed.

File: jung-visualization/src/main/java/edu/uci/ics/jung/visualization/spatial/Spatial.java [hunk 16]
```diff
@@ -0,0 +1,32 @@
+package edu.uci.ics.jung.visualization.spatial;
+
+import com.google.common.collect.Multimap;
+import java.awt.*;
+import java.util.Collection;
+
+/** Created by Tom Nelson on 9/22/17. */
```
Comment (jbduncan) [inline jung-visualization/src/main/java/edu/uci/ics/jung/visualization/spatial/Spatial.java]: Please remove this comment.

File: jung-visualization/src/main/java/edu/uci/ics/jung/visualization/spatial/Spatial.java [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+public interface Spatial<N> {
+  /**
+   * A Multimap of box number to Lists of nodes in that box
+   *
+   * @return
```
Comment (jbduncan) [inline jung-visualization/src/main/java/edu/uci/ics/jung/visualization/spatial/Spatial.java]: This empty `@return` is not needed. Please remove it.

File: jung-visualization/src/main/java/edu/uci/ics/jung/visualization/spatial/SpatialGrid.java [hunk 18]
```diff
@@ -0,0 +1,253 @@
...
+import com.google.common.collect.Multimap;
+import com.google.common.collect.Multimaps;
+import com.google.common.collect.Sets;
+import java.awt.*;
+import java.awt.geom.Point2D;
+import java.util.Collection;
+import java.util.List;
+import java.util.Set;
+
+/** Created by Tom Nelson */
```
Comment (jbduncan) [inline jung-visualization/src/main/java/edu/uci/ics/jung/visualization/spatial/SpatialGrid.java]: Please remove this comment.

File: jung-visualization/src/main/java/edu/uci/ics/jung/visualization/spatial/SpatialGrid.java [hunk 19]
```diff
@@ -0,0 +1,253 @@
...
+    this.horizontalCount = horizontalCount;
+    this.verticalCount = verticalCount;
+    this.boxWidth = size.getWidth() / horizontalCount;
+    this.boxHeight = size.getHeight() / verticalCount;
+  }
+
+  /**
+   * A Multimap of box number to Lists of nodes in that box
+   *
+   * @return
```
Comment (jbduncan) [inline jung-visualization/src/main/java/edu/uci/ics/jung/visualization/spatial/SpatialGrid.java]: Not clear to me that this is useful. Please remove.

File: jung-visualization/src/main/java/edu/uci/ics/jung/visualization/spatial/SpatialGrid.java [hunk 20]
```diff
@@ -0,0 +1,253 @@
...
+    return map;
+  }
+
+  /**
+   * given the box x,y coordinates (not the coordinate system) return the box number (0,0) has box 0
+   * (horizontalCount,horizontalCount) has box horizontalCount*verticalCount - 1
+   *
+   * @param boxX
+   * @param boxY
+   * @return
```
Comment (jbduncan) [inline jung-visualization/src/main/java/edu/uci/ics/jung/visualization/spatial/SpatialGrid.java]: Please fill these out or remove them.

File: jung-visualization/src/main/java/edu/uci/ics/jung/visualization/spatial/SpatialGrid.java [hunk 21]
```diff
@@ -0,0 +1,253 @@
...
+
+  public int getBoxNumber(int[] boxXY) {
+    return getBoxNumber(boxXY[0], boxXY[1]);
+  }
+
+  /**
+   * given a box number, return the x,y coordinates in the grid coordinate system
+   *
+   * @param boxIndex
+   * @return
```
Comment (jbduncan) [inline jung-visualization/src/main/java/edu/uci/ics/jung/visualization/spatial/SpatialGrid.java]: Please fill these out or remove them

Comment (jbduncan) [reply jung-visualization/src/main/java/edu/uci/ics/jung/visualization/spatial/SpatialGrid.java]: This applies to all methods in this class.

File: jung-visualization/src/main/java/edu/uci/ics/jung/visualization/util/LayoutMediator.java [hunk 22]
```diff
@@ -0,0 +1,89 @@
...
+
+import com.google.common.graph.Network;
+import edu.uci.ics.jung.algorithms.layout.Layout;
+import edu.uci.ics.jung.algorithms.util.IterativeContext;
+import java.awt.*;
+import java.awt.geom.Point2D;
+import java.util.Set;
+import java.util.function.Function;
+
+/** Created by Tom Nelson on 9/22/17. */
```
Comment (jbduncan) [inline jung-visualization/src/main/java/edu/uci/ics/jung/visualization/util/LayoutMediator.java]: Please remove this comment and all other "Created by..." comments like this.

File: pom.xml [hunk 23]
```diff
@@ -176,7 +176,7 @@
           <execution>
             <phase>verify</phase>
             <goals>
-              <goal>check</goal>
+              <goal>format</goal>
```
Comment (jbduncan) [inline pom.xml]: Please do NOT change this. If you want to format code, you can run `mvn fmt:format` as described in CONTRIBUTING.md. :)

Comment (jbduncan) [reply pom.xml]: FYI, the reason to not change this is because Travis CI depends on the `mvn fmt` command working this way to check that people have properly formatted their code before check-in.

**Comment by tomnelson** [comment]
Thank you for being thorough with your review of my changes to this
snapshot branch. Did you happen to build and run any of the demos?

I spent quite a bit of time figuring out how visualization became so
broken. I also enjoyed revisiting the codebase and experimenting with
spatial data structures. There is a lot that can be done in that area.
I agree that a lot of cleanup of this snapshot branch is required before
release. I will hold off until I have some idea where the rest of the
project is going.

Nice to meet you!

**Comment by jbduncan** [comment]
> Thank you for being thorough with your review of my changes to this
snapshot branch. Did you happen to build and run any of the demos?
(and when you did, did you see any log messages?)

Hi @tomnelson. I admit that I did not actually run any of the demos. I'll see if I can find and open up my dev machine to run a couple of them, and see if I notice any logging messages that way.

> I spent quite a bit of time figuring out how visualization became so
broken. I also enjoyed revisiting the codebase and experimenting with
spatial data structures. There is a lot that can be done in that area.
I agree that a lot of cleanup of this snapshot branch is required before
release. I will hold off until I have some idea where the rest of the
project is going.

Okey dokey, SGTM!

**Comment by jbduncan** [comment]
@tomnelson Nice to meet you, too! :)

File: jung-visualization/src/main/java/edu/uci/ics/jung/visualization/util/LayoutMediator.java (hunk 22 above)
Comment (tomnelson) [reply jung-visualization/src/main/java/edu/uci/ics/jung/visualization/util/LayoutMediator.java]: Is someone planning to remove all of the author names in comments for the 'legacy' code?

Comment (jbduncan) [reply jung-visualization/src/main/java/edu/uci/ics/jung/visualization/util/LayoutMediator.java]: Good question. :)

Not yet, but I've added it as an item to #101, so it's tracked now.

Comment (jbduncan) [reply jung-visualization/src/main/java/edu/uci/ics/jung/visualization/util/LayoutMediator.java]: There's a couple of reasons. Firstly, I wanted the author(s) removed because I automatically associate comments of this style with auto-generated comments from IDEs, which I personally find unhelpful. Secondly, I just simply hadn't realised before writing my initial comment above that comments like this do already exist in the code base.

I personally advocate either removing author comments of this style, or alternatively rejigging them so they have the following format (which I believe is more standard in the Java ecosystem). But if you'd rather keep author comments as they are for consistency, I'd totally understand.

```java
/**
 * (Class javadoc goes here.)
 *
 * @author Tom Nelson
 */
```

File: pom.xml (hunk 23 above)
Comment (tomnelson) [reply pom.xml]: If you leave it at check, then the build fails. I hope nobody is checking in code when the build failed. If your goal is to have the code formatted, why not have formatting be a part of every build? What am I missing here?

Comment (jrtom) [reply pom.xml]: In theory, no one should be checking in code when the build failed, because we have the Travis check to forestall that (at least for the common.graph and master branches).

I haven't formed a strong opinion on this issue yet, but I think that the point of leaving it as a check is to not require formatting on every build; I can imagine that there might be circumstances in which you might want to leave the formatting until you were done developing.

But maybe there's something I'm missing; @jbduncan, can you elaborate?

**Comment by jrtom** [comment]
Sorry I haven't yet taken a look at this set of changes; I'll try to do that this weekend.

**Comment by tomnelson** [comment]
Unformatted code will make the build fail if you leave it set to 'check'. I
got so tired of git adding and removing the pom or running a separate mvn
format every time i wanted to build, so I thought I would ask what your
real goal is (hopefully not just making contributors add a google code
format plugin to their IDE)

**Comment by tomnelson** [comment]
I wondering if a better path might be separate projects for visualization
and (visualization)samples.

File: pom.xml (hunk 23 above)
Comment (jbduncan) [reply pom.xml]: @jrtom I admit I'm not too clear either on why checking is preferred to formatting at CI-time. I think it's a combination of reasons.

I can think of two reasons ATM. Firstly, it's partially to do with how formatters (google-java-format included) can sometimes contain bugs which can cause code to be formatted weirdly, or to flat-out fail with an error or exception. So, if formatting happened during CI-time, the CI may fail or format things strangely for reasons not to do with the checked-in code, which has the potential to make things confusing.

Secondly, formatters are prone to occasionally formatting code incompletely, thus requiring formatters to be run twice or so, so formatting by CI wouldn't be hermetic, which again is a cause for confusion. I actually experienced this "problem" (if you'd even call it that) with google-java-format when I was working on introducing it to JUNG a good few weeks/months ago; it required me to run `mvn fmt:format` twice before certain source files were formatted completely, strangely.

I hope @nedtwigg, who I collaborate with on [Spotless](https://github.com/diffplug/spotless), can elaborate further on this, because I'm sure an issue was opened in the past on Spotless's issue tracker asking why Spotless's `spotlessCheck` task (similar to `mvn fmt:format` for us) is the default over `spotlessApply` (similar to `mvn fmt:check`), but I can't find that issue yet.

**Comment by tomnelson** [comment]
I would make a default maven profile (that we use) that formats the code,
and a different profile that CI invokes to just do a check of the formatting

File: pom.xml (hunk 23 above)
Comment (jbduncan) [reply pom.xml]: Replying to https://github.com/jrtom/jung/pull/126#issuecomment-333177281:

@tomnelson That sounds good to me, having a default maven profile for formatting code and a CI-specific one for checking code.

Would you be happy to create a new PR for that?

If not, I may have the time and energy in the foreseeable future to investigate Maven profiles and create them myself.

Comment (jrtom) [reply pom.xml]: We definitely wouldn't want the CI check to be formatting code itself; I don't want an automated CI check to be trying to modify the code (I don't even know what that would do, but I feel no strong need to find out :) ).

So I would say that until we have multiple profiles that we should keep it at "check".

**Comment by tomnelson** [comment]
I will add the profile, if there is agreement.
Whoever is admin on the CI server would have to add the profile to the
build that runs there.

**Comment by tomnelson** [comment]
No wonder you guys are so busy! Every time you build, you have to do it
twice! ;)

File: pom.xml (hunk 23 above)
Comment (jbduncan) [reply pom.xml]: Replying to comment https://github.com/jrtom/jung/pull/126#issuecomment-333180321: 

@tomnelson It should be easy to apply the new profile without needing admin permissions. I assume that profiles can be chosen from the command-line? If so, all you'd need to do is change [this line](https://github.com/jrtom/jung/blob/master/.travis.yml#L19) in the Travis config to ensure it runs the right profile.

**Comment by nedtwigg** [comment]
I would discourage you from having the formatter fix changes in CI.  Here's why:
